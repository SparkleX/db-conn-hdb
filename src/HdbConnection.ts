import { Connection, Result, ResultSetMetaData, ResultSetColumnMetadata, SqlType} from "db-conn";
import { Buffer } from "buffer";
import { HdbColumnType } from "./HdbColumnType";
const hdb  = require("hdb");

export class HdbConnection implements Connection {
	private client: any;
	public constructor(client: any) {
		this.client = client;
	}
	public async close(): Promise<void> {
		this.client.end();
		delete this.client;
	}
	private convertColumnMetaData (colMeta: any): ResultSetColumnMetadata {
		let rt: ResultSetColumnMetadata = {};
		rt.name = colMeta.columnName;
		switch(colMeta.dataType) {
			case HdbColumnType.INTEGER :
				rt.type = SqlType.integer;
				break;
			case HdbColumnType.NVARCHAR :
				rt.type = SqlType.varchar;
				break;
			case HdbColumnType.DECIMAL :
				rt.type = SqlType.decimal;
				break;
			case HdbColumnType.TIMESTAMP :
				rt.type = SqlType.timestamp;
				break;
			//default:
				//throw Error(`undefined column type ${colMeta.dataType}`)
		}
		return rt;
	}

	private convertResultSetMetaData (statement: any): ResultSetMetaData {
		const rt:ResultSetMetaData = [];
		for (let i in statement.resultSetMetadata) {
			const srcColMeta = statement.resultSetMetadata[i];
			const tgtColMeta = this.convertColumnMetaData(srcColMeta);
			rt.push(tgtColMeta);
		}
		return rt;

	}
	public async execute(sql: string, params?: object | any[] | undefined): Promise<Result> {
		const thisConn = this;
		if (params === undefined) {
			params = [];
		}
		return new Promise((resolve, reject) => {
			this.client.prepare(sql, function (err: any, statement: any){
				if (err) {
					reject(err);
					return;
				}
				statement.exec(params, function (err: any, rows: any) {
					if (err) {
						reject(err);
						return;
					}
					const rt : Result = {};
					if(isNaN(rows) == false) {
						rt.affectedRows = rows;
					}
					if(Array.isArray(rows)) {
						rt.data = rows;
						rt.metadata = thisConn.convertResultSetMetaData(statement);
						for(const index in rt.data) {
							const obj: any = rt.data[index];
							for(const name in obj) {
								const value = obj[name];
								if (Buffer.isBuffer(value) == false) {
									continue;
								}
								obj[name] = (value as Buffer).toString();
							}
						}
					}
					statement.drop(function(err: any){
						/* istanbul ignore next */
						if (err) {
							reject(err);
							return;
						}
						resolve(rt);
					});					
				});
			});
		});
	}
	public async executeQuery(sql: string, params?: object | any[] | undefined): Promise<object[]> {
		const rt: Result = await this.execute(sql, params);
		if(rt.data === undefined) {
			throw new Error("No data returned");
		}
		return rt.data;
	}
	public async setAutoCommit(autoCommit: boolean): Promise<void> {
		this.client.setAutoCommit(autoCommit);
	}
	public async commit(): Promise<void> {
		const that = this;
		return new Promise((resolve, reject) => {
			that.client.commit(function(err: any){
				if(err) {
					reject(err);
					return;
				}
				resolve();
			});
		});
	}
	public async rollback(): Promise<void> {
		const that = this;
		return new Promise((resolve, reject) => {
			that.client.rollback(function(err: any){
				if(err) {
					reject(err);
					return;
				}
				resolve();
			});
		});
	}
}
