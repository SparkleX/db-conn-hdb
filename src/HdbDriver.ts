import { Driver, Connection, SqlError } from "db-conn";
import { HdbConnection, HdbConnectionConfig } from ".";
import { HdbSqlError } from "./HdbSqlError";
const hdb  = require("hdb");

export class HdbDriver implements Driver {
	connect(config: HdbConnectionConfig): Promise<Connection> {
		const client = hdb.createClient(config);
		return new Promise((resolve, reject) => {
			client.on('error', function (err: any) {
				reject(new HdbSqlError("hdb error", err));
				return ;
			});
			client.connect(function (err: any) {
				if (err) {
					reject(new HdbSqlError("hdb connect failed", err));
					return;
				}
				const conn: Connection = new HdbConnection(client);
				resolve(conn);
			});
		});
	}

}