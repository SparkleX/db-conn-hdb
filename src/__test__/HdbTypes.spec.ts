import { HdbConnection } from "../HdbConnection";
import { Connection } from "db-conn";
import { HdbConnectionConfig } from "../HdbConnectionConfig";
import { HdbDriver } from "../HdbDriver";
import * as jsonfile from "jsonfile"
import { exception } from "console";
const driver = new  HdbDriver();

const config: HdbConnectionConfig = jsonfile.readFileSync("config/test.json");

async function rebuild(conn: Connection):Promise<void> {
	try {
		await conn.execute("drop schema testhdb cascade");
	}catch (e) {		
	}	
	await conn.execute("create schema testhdb");
	await conn.execute("set schema testhdb");
	try {
		await conn.execute(`drop table test`);
	}catch (e) {		
	}	
}

test("Decimal", async () => {
	const conn: Connection = await driver.connect(config);
	await rebuild(conn);
	let rt = await conn.execute(`create table TEST( ID INTEGER not null, CTEXT text, AMOUNT DECIMAL(19,6), times timestamp, primary key ("ID"))`);
	expect(rt).toStrictEqual({});
	rt = await conn.execute(`insert into TEST(ID, AMOUNT, times) values(1, '1234567890123.123456', '2007-02-28 12:45:12.23456')`);
	expect(rt).toStrictEqual({affectedRows:1});

	const result = await conn.execute(`select * from "TEST"`);
	expect(result.metadata).toMatchSnapshot("metadata");
	expect(result.data).toMatchSnapshot("data");
	await conn.close();
});

