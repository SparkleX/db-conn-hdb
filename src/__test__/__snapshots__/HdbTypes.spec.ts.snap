// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Decimal: data 1`] = `
Array [
  Object {
    "AMOUNT": "1234567890123.123456",
    "CTEXT": null,
    "ID": 1,
    "TIMES": "2007-02-28T12:45:12.234",
  },
]
`;

exports[`Decimal: metadata 1`] = `
Array [
  Object {
    "columnName": "ID",
    "columnType": 1,
  },
  Object {
    "columnName": "CTEXT",
    "columnType": 2,
  },
  Object {
    "columnName": "AMOUNT",
    "columnType": 3,
  },
  Object {
    "columnName": "TIMES",
    "columnType": 12,
  },
]
`;
